### PR TITLE
Merge nodeagent/sds work to master

### PR DESCRIPTION
**Pull request by dunjut**
Signed-off-by: dastan <real.duhw@gmail.com>

https://github.com/istio/istio/issues/8009

### CODE REVIEW

**Comment by quanjielin** [comment]
thanks for making the change.

Do you plan to make changes on this after merging to master ? if yes, could you make the change on gcp-identity branch first; The official build will still come from gcp-identity for a while(with various new changes to be added) and there is no way to test change on master now.

File: security/pkg/nodeagent/sds/sdsservice.go [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+	secret, err := s.st.GetSecret(ctx, discReq.Node.Id, spiffeID, token)
+	if err != nil {
+		log.Errorf("Failed to get secret for proxy %q from secret cache: %v", discReq.Node.Id, err)
+		return nil, err
+	}
+
+	return sdsDiscoveryResponse(secret, discReq.Node.Id)
+}
+
+// NotifyProxy send notification to proxy about secret update,
```
Comment (incfly) [inline security/pkg/nodeagent/sds/sdsservice.go]: s/sends

File: security/pkg/nodeagent/sds/sdsservice.go [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+		st: st,
+	}
+}
+
+// register adds the SDS handle to the grpc server
+func (s *sdsservice) register(rpcs *grpc.Server) {
+	sds.RegisterSecretDiscoveryServiceServer(rpcs, s)
+}
+
+func (s *sdsservice) StreamSecrets(stream sds.SecretDiscoveryService_StreamSecretsServer) error {
```
Comment (incfly) [inline security/pkg/nodeagent/sds/sdsservice.go]: comment is needed for public method.

**Comment by myidpt** [comment]
@quanjielin He is volunteering to help merge the work back to master.

**Comment by quanjielin** [comment]
please hold this until basic e2e works in master, any urgent change on nodeagent please check into gcp-identity so we could verify it won't break existing functionality. thanks.

**Comment by dunjut** [comment]
@quanjielin These files come from gcp-identity (92eccb0a5942375a72afadae1ee46ad0a95957f2) without any change, except two comments for public methods and one typo, as suggested by @incfly . :-)

Circle CI tests have already passed, lets wait for those prows.

If I want to change something in code, I should open a PR to gcp-identity then later merge it back to master, as you adviced, right? :P

**Comment by quanjielin** [comment]
@dunjut I understand this PR just copy files between branches without new change. The reason I suggest to hold is because  SDS don't work in master now(pending on envoy+pilot change), so merge don't bring any benefit now, but any SDS new changes(not only yours) need to be checked into two branches, also other ppl without context may change master directly and causes two branches diverge. I appreciate your help, will let you know once it's ready. 

One thing we must finish before merging back to master is break dependency on auth provider in SDS(needs some change in security/pilot); let me know if you're interested and have time to help on that part :)

**Comment by dunjut** [comment]
@quanjielin I would be glad to help :) Though I think the sooner the merge work is done, the easier we make changes later. What do you think? @myidpt

**Comment by myidpt** [comment]
@dunjut let's hold this PR for a while.
@quanjielin please help to assign some task in "break dependency on auth provider in SDS" to dunjut. He is able to help us to accelerate the process :)

**Comment by quanjielin** [comment]
@dunjut let's start with https://github.com/istio/istio/issues/8642

**Comment by quanjielin** [comment]
going to close this PR since sds main part is checked in, I will merge some other changes then try to make e2e works in master.

**Pull request closed without merging**